### PR TITLE
Add class NonMatching::FEValues

### PR DESCRIPTION
**Pull request by simonsticko**
This class makes it easier to assemble cut finite element methods, when the domain is described by a level set function. In the same way as hp::FEValues, NonMatching::FEValues generates dealii::FEValues like objects, but where these contain quadrature rules to integrate over the 3 different regions of the cell that are defined by the sign of the level set function.

Depends on #12024 and #12101.

### CODE REVIEW

File: include/deal.II/non_matching/fe_values.h
Comment (peterrum) [inline include/deal.II/non_matching/fe_values.h:192]: As stated in #12757, I would like to see a version without the collections. Also, I think the 1D quadrature can be simply generated from the higher-dimensional one.

File: include/deal.II/non_matching/fe_values.h
Comment (peterrum) [inline include/deal.II/non_matching/fe_values.h:196]: `VECTOR` -> `VectorType`

File: source/non_matching/fe_values.cc
Comment (peterrum) [inline source/non_matching/fe_values.cc:63]: I guess this is the answer to my question in https://github.com/dealii/dealii/pull/12101#pullrequestreview-754364194.

File: source/non_matching/fe_values.cc
Comment (peterrum) [inline source/non_matching/fe_values.cc:115]: I wouldn't make unit_box a class field.

File: source/non_matching/fe_values.cc
Comment (peterrum) [inline source/non_matching/fe_values.cc:86]: This looks a bit complicated. All you need is a `Function` that can be passed to `QuadratureGenerator`.

**Review by peterrum — COMMENTED** [review]
I like this class. Reminds me a bit of `hp::FEValues` ;)

File: include/deal.II/non_matching/fe_values.h
Comment (simonsticko) [reply include/deal.II/non_matching/fe_values.h:220]: I've added an easier-to-use constructor. It's not exactly what you asked for, but I think it makes sense.

**Comment by simonsticko** [comment]
I've removed the WIP status now that #12024 is merged. I think this is ready for some more comments.

File: include/deal.II/non_matching/fe_values.h
Comment (peterrum) [inline include/deal.II/non_matching/fe_values.h:367]: I think everything below here can go into the source file.

File: include/deal.II/non_matching/fe_values.h
Comment (peterrum) [inline include/deal.II/non_matching/fe_values.h:52]: You need to add `#ifdefs` for doxygen here.

File: include/deal.II/non_matching/fe_values.h
Comment (peterrum) [inline include/deal.II/non_matching/fe_values.h:131]: What was the reason again why we don't use `std::shared_ptr` here?

File: include/deal.II/non_matching/fe_values.h
Comment (peterrum) [inline include/deal.II/non_matching/fe_values.h:183]:
```suggestion
             const Quadrature<dim> &        quadrature,
```
You could use internally `Quadrature::is_tensor_product()` and `Quadrature::get_tensor_basis`.

File: include/deal.II/non_matching/fe_values.h
Comment (peterrum) [inline include/deal.II/non_matching/fe_values.h:232]:
```suggestion
     * get_outside_fe_values(), or
```

File: include/deal.II/non_matching/fe_values.h
Comment (peterrum) [inline include/deal.II/non_matching/fe_values.h:360]:
```suggestion
      reference_space_level_set;
```

File: source/non_matching/fe_values.cc
Comment (peterrum) [inline source/non_matching/fe_values.cc:3]:
```suggestion
// Copyright (C) 2021 by the deal.II authors
```

File: source/non_matching/fe_values.cc
Comment (peterrum) [inline source/non_matching/fe_values.cc:81]:
```suggestion
        element = &dof_handler->get_fe();
```

File: source/non_matching/fe_values.cc
Comment (peterrum) [inline source/non_matching/fe_values.cc:88]:
```suggestion
        for (unsigned int i = 0; i < local_dof_indices.size(); ++i)
```

File: source/non_matching/fe_values.cc
Comment (peterrum) [inline source/non_matching/fe_values.cc:91]: Why not use a dummy `AffineConstraints` object?

File: source/non_matching/fe_values.cc
Comment (peterrum) [inline source/non_matching/fe_values.cc:106]: We normally prefer if functions return bool and the calling function is asserting. You could even create an own excaption.

**Review by peterrum — COMMENTED** [review]
Looks good to me! I left a few minor comments.

File: include/deal.II/non_matching/fe_values.h
Comment (simonsticko) [reply include/deal.II/non_matching/fe_values.h:131]: 1. `optional` communicates more clearly that the returned object may not contain a value.
2. On non-cut cells, we can set the `optional` to the reference we get from `hp::FEValues::get_present_fe_values()`. A `shared_ptr`'s destructor would try to delete that object.

File: include/deal.II/non_matching/fe_values.h
Comment (simonsticko) [reply include/deal.II/non_matching/fe_values.h:183]: Yes, but wouldn't that be more unclear? One would first have to assert that the incoming `Quadrature<dim>` is a tensor product and that each `Quadrature<1>` in the basis is the same, otherwise we have to choose which component to build the immersed quadrature rules from.

File: source/non_matching/fe_values.cc
Comment (simonsticko) [reply source/non_matching/fe_values.cc:81]: I don't think this will not work, different cells might use different elements. This gives the first element in the `FECollection`.

File: include/deal.II/non_matching/fe_values.h
Comment (simonsticko) [reply include/deal.II/non_matching/fe_values.h:370]: I had a test for `RefSpaceFEFieldFunction`, but I had forgotten to add it. That is why this is in the header. We can either keep this in the header or remove the test. What do you think is best?

**Comment by simonsticko** [comment]
I got some weird doxygen warnings that I don't know how to work around:
```
[100%] Generating documentation via doxygen.
-- Documentation is available at /home/runner/work/dealii/dealii/build/doc/doxygen/deal.II/index.html
[100%] Built target doxygen
[100%] Built target documentation
/home/runner/work/dealii/dealii/source/fe/fe_values.cc:4160: warning: no uniquely matching class member found for 
  template < dim, spacedim >
  const unsigned int FEValues< dim, spacedim >::integral_dimension()

/home/runner/work/dealii/dealii/source/fe/fe_values.cc:4165: warning: no uniquely matching class member found for 
  template < dim, spacedim >
  FEValues< dim, spacedim >::FEValues(const Mapping< dim, spacedim > &mapping, const FiniteElement< dim, spacedim > &fe, const Quadrature< dim > &q, const UpdateFlags update_flags)
```

File: include/deal.II/non_matching/fe_values.h
Comment (peterrum) [reply include/deal.II/non_matching/fe_values.h:134]: Is it possible to use https://github.com/dealii/dealii/blob/master/include/deal.II/base/std_cxx17/optional.h?

File: include/deal.II/non_matching/fe_values.h
Comment (peterrum) [reply include/deal.II/non_matching/fe_values.h:186]: The reason why I would prefer `Quadrature<dim>` is that `FEValues` takes one of these, i.e., there will be probably an instance of this lying around. What opinion have others?

File: include/deal.II/non_matching/fe_values.h
Comment (peterrum) [reply include/deal.II/non_matching/fe_values.h:370]: I would remove the test and move the content to the source file. `RefSpaceFEFieldFunction` indirectly via `NonMatching::FEValues`.

File: source/non_matching/fe_values.cc
Comment (peterrum) [reply source/non_matching/fe_values.cc:88]: True!

**Comment by peterrum** [comment]
The warning looks odd? Has anyone else observed such issues?

File: include/deal.II/non_matching/fe_values.h
Comment (simonsticko) [reply include/deal.II/non_matching/fe_values.h:134]: Unfortunately not, `boost::optional<T>` allows `T` to be a reference, `std::optional<T>` does not, and I want to avoid copying the `FEValues` object.

File: source/non_matching/fe_values.cc
Comment (simonsticko) [reply source/non_matching/fe_values.cc]: I tried this:
```c++

const AffineConstraints<> dummy_constraints;
dummy_constraints.get_dof_values(*global_dof_values,
                                 local_dof_indices.begin(),
                                 local_dof_values.begin(),
                                 local_dof_values.end());
```
But it doesn't seem to work for `EpetraWrappers::Vector`

```
2021-11-19T15:40:38.4878272Z /home/runner/work/dealii/dealii/include/deal.II/lac/affine_constraints.h: In instantiation of ‘void dealii::AffineConstraints<number>::get_dof_values(const VectorType&, ForwardIteratorInd, ForwardIteratorVec, ForwardIteratorVec) const [with ForwardIteratorVec = __gnu_cxx::__normal_iterator<double*, std::vector<double, std::allocator<double> > >; ForwardIteratorInd = __gnu_cxx::__normal_iterator<unsigned int*, std::vector<unsigned int> >; VectorType = dealii::LinearAlgebra::EpetraWrappers::Vector; number = double]’:
2021-11-19T15:40:38.4945499Z /home/runner/work/dealii/dealii/source/non_matching/fe_values.cc:96:9:   required from ‘void dealii::NonMatching::internal::FEValuesImplementation::RefSpaceFEFieldFunction<dim, VectorType>::set_active_cell(const typename dealii::Triangulation<dim>::active_cell_iterator&) [with int dim = 1; VectorType = dealii::LinearAlgebra::EpetraWrappers::Vector; typename dealii::Triangulation<dim>::active_cell_iterator = dealii::TriaActiveIterator<dealii::CellAccessor<1, 1> >]’
2021-11-19T15:40:38.4948301Z /home/runner/work/dealii/dealii/build/source/non_matching/fe_values.inst:991:2:   required from here
2021-11-19T15:40:38.4950239Z /home/runner/work/dealii/dealii/include/deal.II/lac/affine_constraints.h:2337:44: error: no match for call to ‘(const dealii::LinearAlgebra::EpetraWrappers::Vector) (unsigned int&)’
2021-11-19T15:40:38.4951503Z          *local_vector_begin = global_vector(*local_indices_begin);
2021-11-19T15:40:38.4952226Z                                ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
2021-11-19T15:40:38.4953966Z /home/runner/work/dealii/dealii/include/deal.II/lac/affine_constraints.h:2344:36: error: no match for call to ‘(const dealii::LinearAlgebra::EpetraWrappers::Vector) (const unsigned int&)’
2021-11-19T15:40:38.4955245Z              value += (global_vector(position.entries[j].first) *
2021-11-19T15:40:38.4955955Z                        ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: include/deal.II/non_matching/fe_values.h
Comment (peterrum) [reply include/deal.II/non_matching/fe_values.h:134]: I agree that `std::optional<T &>` would be the right way and that it does not work is a pity. Our policy is normally that we tend to prefer the standardized  functionalities from `std` once they are available (and if they have the same functionalities). In this case, 99% of the functionality is the same (between `std::optional` and `std::boost`) but one feature, which we find useful is missing. Are there any options what to do in such a situation? @bangerth What is your opinion in this regards?

Comment (simonsticko) [reply include/deal.II/non_matching/fe_values.h:134]: Thinking a bit more about this I think I could return
`const std::optional<FEValues<dim>> &`
If I avoid using `hp::FEValues` internally. Would that be acceptable?

Comment (peterrum) [reply include/deal.II/non_matching/fe_values.h:134]: I guess your idea is to replace the following lines:

https://github.com/dealii/dealii/blob/542b0ef7d8b58225591b6e22ae0251671fe06350/source/non_matching/fe_values.cc#L355-L356

https://github.com/dealii/dealii/blob/542b0ef7d8b58225591b6e22ae0251671fe06350/source/non_matching/fe_values.cc#L377-L378

with the creation of new `FEValues` instances? I don't think this is a good idea, since you make by that the non-cut case slower (creating new `FEValues` instances is not that cheap since all the memory for internal buffers has to allocated).

Comment (simonsticko) [reply include/deal.II/non_matching/fe_values.h:134]: No, instead of `hp::FEValues` as a member variable, I would need to keep the `dealii::FEValues` objects that I need as a member variable:
```c++
std::vector<std::optional<dealii::FEValues<dim>>> fe_values_full_quadrature;
```
Then I can return a reference to one of these `optional`s based on `active_fe_index`.

Comment (peterrum) [reply include/deal.II/non_matching/fe_values.h:134]: I see. I din't think about this. That is not elegant but definitely an option if we don't find a better solution. Let's wait a couple of days for feedback from others!

Comment (masterleinad) [reply include/deal.II/non_matching/fe_values.h:134]: What about `std::optional<std::reference_wrapper<...>>`?

Comment (simonsticko) [reply include/deal.II/non_matching/fe_values.h:134]: That would work, but then one would have to write `.get()` everywhere when assembling:
```c++
  std::optional<std::reference_wrapper<FEValues<dim>>> fe_values;
  // fe_values->JxW(0); //Doesn't compile
  fe_values->get().JxW(0); // Compiles
```

Comment (peterrum) [reply include/deal.II/non_matching/fe_values.h:134]: Till now, I think `std::vector<std::optional<dealii::FEValues<dim>>> fe_values_full_quadrature;` would be the best solution. My question would be if the vector is enough? `hp::FEValues` constructs a 3D table for each mapping-, FE-, quadrature-object. Do we need this full flexibility here or is it enough that only on mapping-, FE-, quadrature-object corresponds?

Comment (simonsticko) [reply include/deal.II/non_matching/fe_values.h:134]: I think it seems like the best solution too.  I think the full table is a bit overkill. If someone needs it in the future, the class can be extended later.

Comment (kronbichler) [reply include/deal.II/non_matching/fe_values.h:134]: I also agree here, I would keep it as `vector<optional<FEValues>>`.

Comment (peterrum) [reply include/deal.II/non_matching/fe_values.h:134]: @simonsticko Could change the code so that it uses `std::vector<std::optional<dealii::FEValues<dim>>>` and add a `#ifndef DOXYGEN` to the problematic function. I think once these two points are done, the PR is ready to go!

Comment (simonsticko) [reply include/deal.II/non_matching/fe_values.h:134]: Okey, but I don't understand how to fix the doxygen problem. It seems to come from `fe/fe_values.cc`.

Comment (peterrum) [reply include/deal.II/non_matching/fe_values.h:134]: Add a big ifndef around everything.

File: source/non_matching/fe_values.cc
Comment (peterrum) [inline source/non_matching/fe_values.cc:96]: I think you also need to revert this change. `linux debug parallel simplex` test indicates that this function call does not work for `dealii::LinearAlgebra::EpetraWrappers::Vector` :(

**Review by peterrum — APPROVED** [review]
@simonsticko  I am OK with the current status. Thank you! Let's wait for the review of one more person!

**Comment by peterrum** [comment]
@simonsticko  Could you push again. GitHub had server issues.

File: include/deal.II/non_matching/fe_values.h
Comment (kronbichler) [inline include/deal.II/non_matching/fe_values.h:66]:
```suggestion
   * how the FEValues objects, which are created by NonMatching::FEValues,
```

File: include/deal.II/non_matching/fe_values.h
Comment (kronbichler) [inline include/deal.II/non_matching/fe_values.h:94]:
```suggestion
   * This class is intended to facilitate assembling in immersed (in the sense of
```

File: include/deal.II/non_matching/fe_values.h
Comment (kronbichler) [inline include/deal.II/non_matching/fe_values.h:108]:
```suggestion
   * discrete level set function describes by a DoFHandler and a Vector. When the reinit()
```

File: include/deal.II/non_matching/fe_values.h
Comment (kronbichler) [inline include/deal.II/non_matching/fe_values.h:109]:
```suggestion
   * function is called, the QuadratureGenerator will be called in the background
```

File: include/deal.II/non_matching/fe_values.h
Comment (kronbichler) [inline include/deal.II/non_matching/fe_values.h:111]:
```suggestion
   * dealii::FEValues objects for the inside/outside regions and an
```

File: include/deal.II/non_matching/fe_values.h
Comment (kronbichler) [inline include/deal.II/non_matching/fe_values.h:119]:
```suggestion
   * points. This can, for example, happen if the relative size of the cut is similar as the floating point accuracy.
```

File: include/deal.II/non_matching/fe_values.h
Comment (kronbichler) [inline include/deal.II/non_matching/fe_values.h:120]:
```suggestion
   * Since the FEValues-like objects are not allowed to contain 0 points, the
```

File: include/deal.II/non_matching/fe_values.h
Comment (kronbichler) [inline include/deal.II/non_matching/fe_values.h:246]:
```suggestion
     * @note If the quadrature rule over the region is empty, e.g. because the
     *  cell is completely located in the outside domain, the returned
     * optional will not contain a value.
```

File: include/deal.II/non_matching/fe_values.h
Comment (kronbichler) [inline include/deal.II/non_matching/fe_values.h:322]:
```suggestion
     * When LocationToLevelSet of the cell is INSIDE (and we do not need
```

File: include/deal.II/non_matching/fe_values.h
Comment (kronbichler) [inline include/deal.II/non_matching/fe_values.h:324]:
```suggestion
     * object in this container corresponding to the cell's active_fe_index.
```

File: include/deal.II/non_matching/fe_values.h
Comment (kronbichler) [inline include/deal.II/non_matching/fe_values.h:334]:
```suggestion
     * this object contains an dealii::FEValues object created with a quadrature rule over
```

File: include/deal.II/non_matching/fe_values.h
Comment (kronbichler) [inline include/deal.II/non_matching/fe_values.h:338]:
```suggestion
     * When LocationToLevelSet of the cell is OUTSIDE (and we do not need
```

File: include/deal.II/non_matching/fe_values.h
Comment (kronbichler) [inline include/deal.II/non_matching/fe_values.h:327]:
```suggestion
     * This container is a std::deque, which is compatible with the `FEValues` class
     * that does not have a copy-constructor.
```
(Use the same in the places below.)

File: include/deal.II/non_matching/fe_values.h
Comment (kronbichler) [inline include/deal.II/non_matching/fe_values.h:349]:
```suggestion
     * FEValues object created with a quadrature formula integrating over the inside
```
(same below)

File: tests/non_matching/fe_values.cc
Comment (kronbichler) [inline tests/non_matching/fe_values.cc:44]: Can't you use `cell == expected`?

File: source/non_matching/fe_values.cc
Comment (kronbichler) [inline source/non_matching/fe_values.cc:314]: I think we most commonly spell out the full word for this function name, `initialize`, so I would lean towards that name unless we have reasons to stay with this abbreviation (as opposed to `reinit`).

**Review by kronbichler — APPROVED** [review]
Thanks for this work and the very extensive comments. I have a number of minor suggestions regarding the documentation, but I think the overall state is very good already.

File: include/deal.II/non_matching/fe_values.h
Comment (kronbichler) [reply include/deal.II/non_matching/fe_values.h:184]: I think in >90% of cases one creates a quadrature right at the place of setting up `FEValues`, so in that sense I do not believe that this should be the deciding factor. Indeed, the 1D quadrature formula signals clearly to the user that we need a very specific tensor product, due to the nature in the construction of the cut quadrature formula at this point. For me, the criterion would be whether we can think of an extension of the interface to a different class of algorithms to define cut quadrature formulas and if we can anticipate anyone interested in contributing these in the mid term (say in the next few years). @bergbauer what is your opinion here?

If we go with the dim-dimensional quadrature formula, we need to write more documentation, and we need to find good ways to compare quadrature objects as suggested by @simonsticko - we would need to inspect the `get_tensor_basis` objects. At least it is not hopeless to do so, so I would make it dependent on the possibility for extension.

**Comment by peterrum** [comment]
@simonsticko Could you add the ifdefs? Once `indent` passes, the PR should be ready to be merged!

File: source/fe/fe_values.cc
Comment (peterrum) [inline source/fe/fe_values.cc:5007]: You might need to ifdef the whole content of this file...

**Review by peterrum — APPROVED** [review]
OK. `indent` is happy. I let @kronbichler do the final call!

**Review by kronbichler — APPROVED** [review]

**Comment by kronbichler** [comment]
Thanks for this big effort, @simonsticko ! This is a nice feature to have included in deal.II.

**Pull request merged**
